src/metabase/util/ssh.clj

```diff
@@ -0,0 +1,95 @@
+(ns metabase.util.ssh
+  (:require [clojure.tools.logging :as log]
+            [metabase.util :as u])
+  (:import com.jcraft.jsch.JSch))
+
+(def default-ssh-timeout 30000)
+
+(defn start-ssh-tunnel
+  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass host port]}]
+  (let [connection (doto (.getSession (new com.jcraft.jsch.JSch) tunnel-user tunnel-host tunnel-port)
+                     (.setPassword tunnel-pass)
+                     (.setConfig "StrictHostKeyChecking" "no")
+                     (.connect default-ssh-timeout)
+                     (.setPortForwardingL 0 host port))
+        input-port (some-> (.getPortForwardingL connection)
+                           first
+                           (clojure.string/split #":")
+                           first
+                           (Integer.))]
+    (assert (number? input-port))
+    (log/info (u/format-color 'cyan "creating ssh tunnel %s@%s:%s -L %s:%s:%s" tunnel-user tunnel-host tunnel-port input-port host port))
+    [connection input-port]))
+
+(def ssh-tunnel-preferences
+  [{:name         "tunnel-enabled"
+    :display-name "Use SSH tunnel"
+    :placeholder  "Enable this ssh tunnel?"
+    :type         :boolean
+    :default      false}
+   {:name         "tunnel-host"
+    :display-name "SSH tunnel host"
+    :placeholder  "What hostname do you use to connect to the SSH tunnel?"}
+   {:name         "tunnel-port"
+    :display-name "SSH tunnel port"
+    :type         :integer
+    :default      22}
+   {:name         "tunnel-user"
+    :display-name "SSH tunnel username"
+    :placeholder  "What username do you use to login to the SSH tunnel?"}
+   {:name         "tunnel-pass"
+    :display-name "SSH tunnel password"
+    :type         :password
+    :placeholder  "******"}
+   #_{:name         "tunnel-private-key"
+    :display-name "SSH private key to connect to the tunnel"
+    :type         :string
+    :placeholder  "Paste the contents of an ssh private key here"}
+   #_{:name         "tunnel-private-key-file-name"
+    :display-name "Path on the Metabase server to a SSH private key file to connect to the tunnel"
+    :type         :string
+    :placeholder  "/home/YOUR-USERNAME/.ssh/id_rsa"}])
+
+(defn with-tunnel-config
+  "Add preferences for ssh tunnels to a drivers :details-fields"
+  [driver-config-map]
+  (assoc driver-config-map :details-fields
+    (constantly (concat ((:details-fields driver-config-map))
+                        ssh-tunnel-preferences))))
```

Comment:
That composition pattern is used in lots of other places like ring-middleware. It can we written as:

    (-> original-thing
        with-ssh-details
        with-somethings-else
        with-more-stuff

the initial version used concat here and was moved up to slightly reduce code duplication. This method allows other fields to be modified as well, which is not really required to the drivers that currently support ssh-tunnels :-/ I can also change it back if this is preferred. In general I lean towards separating and then composing over mixing in